aip/0160.md

```diff
@@ -0,0 +1,173 @@
+literals and logical operators.
+
+Because filters are accepted as query strings, type conversion takes place to
+translate the string to the appropriate strongly-typed value:
+
+- Enums expect the enum's string representation (case-sensitive).
+- Booleans expect `true` and `false` literal values.
+- Numbers expect the standard integer or float representations. For floats,
+  exponents are supported (e.g. `2.997e9`).
+- [Durations][] expect a numeric representation followed by an `s` suffix (for
+  seconds). Examples: `20s`, `1.2s`.
+- [Timestamps][] expect an [RFC-3339][] formatted string (e.g.
+  `2012-04-21T11:30:00-04:00`). UTC offsets are supported.
+
+**Warning:** The identifiers `true`, `false`, and `null` only carry intrinsic
+meaning when used in the context of a typed field reference.
+
+Additionally, when comparing strings for equality, the `*` character denotes
+wildcard syntax; for example, `a = "*.foo"` is true if `a` _ends with_ ".foo".
+
+### Traversal operator
+
+Filtering implementations **must** provide the `.` operator, which indicates
+traversal through a message, map, or struct.
+
+| Example         | Meaning                                               |
+| --------------- | ----------------------------------------------------- |
+| `a.b = true`    | True if `a` has a boolean `b` field that is true.     |
+| `a.b > 42`      | True if `a` has a numeric `b` field that is above 42. |
+| `a.b.c = "foo"` | True if `a.b` has a string `c` field that is "foo".   |
+
+**Important:** The `.` operator **must not** be used to traverse through a
+repeated field or list, except for specific use with the `:` operator.
+
+### Has Operator
+
+Filtering implementations **must** provide the `:` operator, which means "has".
+It is usable with collections (repeated fields or maps) as well as messages,
+and behaves slightly differently in each case.
+
+Repeated fields query to see if the repeated structure contains a matching
+element:
+
+| Example    | Meaning                                                     |
+| ---------- | ----------------------------------------------------------- |
+| `r:42`     | True if `r` contains 42.                                    |
+| `r.foo:42` | True if `r` contains an element `e` such that `e.foo = 42`. |
+
+**Important:** Filters can not query a _specific_ element on a repeated field
+for a value. For example, `e.0.foo = 42` and `e[0].foo = 42` are **not** valid
+filters.
+
+Maps, structs, messages can query either for the presence of a field in the map
+or a specific value:
+
+| Example    | Meaning                             |
+| ---------- | ----------------------------------- |
+| `m:foo`    | True if `m` contains the key "foo". |
+| `m.foo:*`  | True if `m` contains the key "foo". |
+| `m.foo:42` | True if `m.foo` is 42.              |
+
+There are two slight distinctions when parsing messages:
+
+- When traversing messages, a field is only considered to be present if it has
+  a non-default value.
```

Comment:
This is tricky because the comment it's true for proto3, but json and proto2 objects would return if the field is set to anything.

Reply:
I think I am okay with not calling that out, since we ostensibly require proto3 and I expect discussing it would be confusing. Does that seem reasonable to you?

Reply:
Okay, great. Works for me